ast/src/main/java/com/graphicsfuzz/common/util/UpgradeShadingLanguageVersion.java

```diff
@@ -92,6 +173,73 @@ public void visitVariableIdentifierExpr(VariableIdentifierExpr variableIdentifie
     }
   }
 
+  @Override
+  public void visitVariablesDeclaration(VariablesDeclaration variablesDeclaration) {
+    final Type baseType = variablesDeclaration.getBaseType();
+    if (baseType.hasQualifier(TypeQualifier.VARYING)) {
+      ((QualifiedType) baseType).replaceQualifier(TypeQualifier.VARYING,
+          TypeQualifier.SHADER_INPUT);
+    }
+    visit(baseType);
+    for (VariableDeclInfo vdi : variablesDeclaration.getDeclInfos()) {
+      if (vdi.hasArrayInfo()) {
+        visit(vdi.getArrayInfo().getSizeExpr());
+      } else if (baseType instanceof ArrayType) {
+        visit(((ArrayType) baseType).getArrayInfo().getSizeExpr());
+      }
+      if (vdi.hasInitializer()) {
+        visit(vdi.getInitializer());
+        // If not constant and has initializer, move the initializer to main.
+        if (!baseType.hasQualifier(TypeQualifier.CONST)) {
+          globals.add(new ExprStmt(new BinaryExpr(new VariableIdentifierExpr(vdi.getName()),
+              vdi.getInitializer().getExpr(), BinOp.ASSIGN)));
+          vdi.setInitializer(null);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void visitFunctionCallExpr(FunctionCallExpr functionCallExpr) {
```

Comment:
Can you add brief comments for each of the four chunks of cases to explain which legacy feature is being mapped to which modern feature?

Reply:
I added a longer comment block explaining the overview.